framework/include/physics/PhysicsBase.h

```diff
@@ -0,0 +1,248 @@
+  bool nonLinearVariableExists(const VariableName & var_name, bool error_if_aux) const;
+
+  /// Add a new required task for all physics deriving from this class
+  /// NOTE: This does not register the task, you still need to call registerMooseAction
+  void addRequiredPhysicsTask(const std::string & task) { _required_tasks.insert(task); }
+
+  /// System number for the systems owning the variables
+  const unsigned int _sys_number;
+
+  /// Whether to output additional information
+  const bool _verbose;
+
+  /// Whether to add a default preconditioning.
+  /// The implementation of the default is defined by the derived class
+  const MooseEnum & _preconditioning;
+
+  /// Keep track of the subdomains the Physics is defined on
+  std::vector<SubdomainName> _blocks;
+
+private:
+  /// Gathers additional parameters for the relationship managers from the Physics
+  /// then calls the parent Action::addRelationshipManagers with those parameters
+  using Action::addRelationshipManagers;
+  void addRelationshipManagers(Moose::RelationshipManagerType input_rm_type) override;
+
+  /// Process some parameters that require the problem to be created. Executed on init_physics
+  void initializePhysics();
+
+  /// The default implementation of these routines will do nothing as we do not expect all Physics
+  /// to be defining an object of every type
+  virtual void addNonlinearVariables() {}
+  virtual void addAuxiliaryVariables() {}
+  virtual void addInitialConditions() {}
+  virtual void addFEKernels() {}
+  virtual void addFVKernels() {}
+  virtual void addNodalKernels() {}
+  virtual void addDiracKernels() {}
+  virtual void addDGKernels() {}
+  virtual void addScalarKernels() {}
+  virtual void addInterfaceKernels() {}
+  virtual void addFVInterfaceKernels() {}
+  virtual void addFEBCs() {}
+  virtual void addFVBCs() {}
+  virtual void addNodalBCs() {}
+  virtual void addPeriodicBCs() {}
+  virtual void addFunctions() {}
+  virtual void addAuxiliaryKernels() {}
+  virtual void addMaterials() {}
+  virtual void addFunctorMaterials() {}
+  virtual void addUserObjects() {}
+  virtual void addPostprocessors() {}
+  virtual void addVectorPostprocessors() {}
+  virtual void addReporters() {}
+  virtual void addOutputs() {}
+  virtual void addPreconditioning() {}
+  virtual void addExecutioner() {}
+  virtual void addExecutors() {}
+
+  /// Check the list of required tasks for missing tasks
+  void checkRequiredTasks() const;
+
+  /// Whether the physics is to be solved as a transient. It can be advantageous to solve
+  /// some physics directly to steady state
+  MooseEnum _is_transient;
+
+  /// Vector of the nonlinear variables in the Physics
+  std::vector<VariableName> _nl_var_names;
+
+  /// Dimension of the physics, which we expect for now to be the dimension of the mesh
+  /// NOTE: this is not known at construction time, only after initializePhysics which is a huge bummer
+  unsigned int _dim;
```

Comment:
```suggestion
  unsigned int _dim = libMesh::invalid_uint;
```
or do something like this in the constructor so that your assertion in `dimension()` works

Reply:
yeah this was probably done in the original PR. Forgot to move it